Document how to declare new modules

blinkseb opened the pull request:
This starts the documentation on how to add new modules. Only the basic stuff is introduced for the moment, which should already help users to get started if they want to add something. The documentation is only valid if https://github.com/MoMEMta/MoMEMta/pull/151 is merged (the only impacted part of this documentation is the `Sticky` part, everything else is valid if we don't merge the PR).

I plan to document advanced usage too (inputs and outputs, etc.), not sure when, I'll try today! In the mean time, don't hesitate if you have feedback on what it's already written :smiley:

Commit:
Document how to declare new modules

swertz commented on docs/in-depth/new-module.md:
```diff
@@ -0,0 +1,63 @@
+
+## Loading the module
+
+So, you have successfully defining a new module and registered it with the system. In order to use it during the computation of the weight, you first have to _load_ the shared library where your module is.
```
defined

swertz replied:
Would it be obvious to users how to compile the module into a library?

blinkseb replied:
Not really I know, but that fell more into "How i compile C++" than "How I use MoMEMta", I'm not sure how to document that

swertz replied:
Actually we could imagine having a script taking care of it?

blinkseb replied:
Hm I'm not sure, it really depends how the user wants to build its code. You can use plain makefiles, cmake, autoconf, etc.

swertz commented:
Thanks a lot :+1:

We can wait a bit before merging so if you have more knock yourself out ;)

Maybe this page could be updated (top) to link the new page: https://github.com/MoMEMta/momemta.github.io/blob/website/docs/in-depth/calling-momemta.md

swertz commented on docs/in-depth/new-module.md:
`````diff
@@ -0,0 +1,192 @@
+
+````
+
+The first addition grab an input of type `double` from the memory pool, expected to be produced by a module named `module` and called `output` (remember that the _output_ of a module becomes the _input_ of another module). The second addition print the actual value of the input at each integration step. In order to access the value of the input, you must dereference the variable (`Value` actually ...
`````
`grabs` and `prints`

swertz commented on docs/in-depth/new-module.md:
````diff
@@ -0,0 +1,192 @@
+};
+```
+
+First addition register an output named `my_output` for this module, of type `double`. The second addition set the value of the output to be twice the input value.
````
same here ;)

swertz commented on docs/in-depth/new-module.md:
```diff
@@ -0,0 +1,192 @@
+!!! note
+    Only a subset of types are supported by `ParameterSet`. You can only store and retrieve `int64_t` (integers), `double` (floating point numbers), `bool`, `std::string`, `InputTag`, `ParameterSet`, as well as `std::vector` of any of the previous types.
+
+We will modify our module to remove the hardcoded `InputTag` and except the user to pass it as a parameter:
```
`except` -> `require`?

Commit:
[Rebase] Fix typos

Commit:
[Squash] More doc

swertz commented:
Thanks a lot for the additions! :+1: It's looking really good, I think there is enough for now or do you have something else in mind?! 
The only thing that's perhaps less clear is the difference of `[Optional]Input` with and without 's'?

blinkseb commented:
I think I documented everything I wanted :) Indeed the `OptionalInput(s)` is a bit tricky, maybe it's better to just document the `Input` method? I was kind of hopping the naming would be self-explanatory but:
  - `Input` expect a single `InputTag`
  - `Inputs` expect a vector of `InputTag`
  - `Optional` mark the input as... optional :smile:

swertz commented on docs/in-depth/new-module.md:
```diff
@@ -0,0 +1,273 @@
+
+##### Supported types
+
+The following types are supported when registering a new attribute:
```
There's also the Paths but maybe that's too technical to mention here?

blinkseb replied:
Indeed you're right, but Paths are not intended to be used by others modules than the Looper (it won't work at all) so let's keep that under the hood for the moment 😄

swertz commented:
Yes it's probably fine; if a user wants to write a module he'll probably end up checking how it's done for  other modules anyway ;)

blinkseb commented:
Indeed that would make sense! I think I'll add a sentence at the end with a link to the `modules` directory of the MoMEMta repo to help the users browse available modules.

Commit:
[Squash] Address comments

blinkseb commented:
I updated the doc, removing the `Inputs` and `Optional...` description. I also added at the end a link to the modules source code.

On my side I think it's ready, but as I said, it's only valid if the associate PR In MoMEMta is merged, so we can probably wait a bit before merging this one in case there're changes to be done

swertz commented:
Thanks! I agree, we can wait some more, I'm still going through the PR...

Commit:
Document new  parameter

swertz commented:
This can now be merged since MoMEMta has been updated. Thanks again! :+1:
